Review Subscription component for Purchase Flow

alex-deaconu opened the pull request.

alex-deaconu commented:
Staged at https://apps-stage-3.risevision.com/

Shows In App Payment Process when `origin != Rise`. On the Test server, it will always show because that field is not available (we haven't set the latest Core version default). Otherwise, a specific Testing company will be needed for testing.

To note, the CSS needs to be reviewed by @ghislaineguerin (I haven't made any changes there).

Most of the prototype code is still there, we will be cleaning it out as we implement the various checkout pages. So far this covers only the first page.

@Rise-Vision/apps please review. Thanks!

fjvallarino commented on src/js/components/purchase-flow/dtv-review-subscription.js:
```diff
@@ -0,0 +1,30 @@
+          };
+
+          $scope.decrementLicenses = function () {
+            $scope.plan.additionalDisplayLicenses--;
```
@alex-deaconu should we validate ```additionalDisplayLicenses``` is greater than zero?

fjvallarino commented on src/templates/components/purchase-flow/checkout-review-subscription.html:
```diff
@@ -0,0 +1,73 @@
+            <input type="radio" name="billingPeriod" id="yearlyBilling" ng-value="false" ng-model="plan.monthlyPrices">
+            <span translate translate-values="{ yearlyPrice: getYearlyPrice() }">common-header.purchase.review-subscription.billed-yearly</span>
+          </label>
+          <label class="highlight" translate translate-values="{ saveYearly: plan.yearly.save }">common-header.purchase.review-subscription.save-yearly</label>
```
@alex-deaconu should this show an amount related to the number of licenses? Currently it shows "Save $10!" independently of the number of licenses

alex-deaconu replied:
@fjvallarino I'm not aware of any discounts for purchasing Per Display Licenses yearly vs. monthly. It's still $10 * 12 for the year so there is no savings. But, maybe I'm mistaken... Thanks.

alex-deaconu commented:
@fjvallarino I've added the validation suggested in the comment.

Also as per @ezequielc 's review I've added validation to the Additional Licenses field:
- allow only numbers (field type=number)
- range between 0-999
- required
- no decimals

Similar validation is added to the directive for various functions. Also added blocks for page navigation if forms are invalid.

Please review latest commit. Thanks!

alex-deaconu commented:
@RV-Adrian I'm assuming this breaks the styling because of the differences in class names. Wondering if you can make a separate PR for your commits? Thanks.

olegrise commented:
@alex-deaconu what is preventing us from setting the latest Core version to be default? Do you know?

alex-deaconu commented:
@olegrise nothing, it's just that once that version is default (and origin='Rise' will be true) the In App payment process would only be accessible from Companies where origin='Chargebee'. Thanks.

RV-Adrian commented:
@alex-deaconu Yes it does, and for that reason I'm suggesting in another comment we skip BEM notation for now and move forward with that style in mind. Will have that resolved ASAP, and then should have no impact on your functionality (ie. just organizing remains)

Sorry, realized after I pushed that it'd be cleaner on another branch. Next time!

olegrise commented on src/js/components/purchase-flow/svc-purchase-factory.js:
```diff
@@ -0,0 +1,29 @@
+      function ($modal, $templateCache) {
+        var _factory = {};
+
+        _factory.showPurchaseModal = function (plan, monthlyPrices) {
```
@alex-deaconu the name `monthlyPrices` is confusing. I was wondering why you are passing monthly prices and not passing the yearly prices. It took me some time to realized that this is a boolean flag.  I think renaming it to `isMonthlyPrices` or even `isMonthly` would clear the confusion.

alex-deaconu replied:
@olegrise good point, will update. Thanks!

alex-deaconu replied:
Re-named variable across the board. Please see latest commit. Thanks.

alex-deaconu commented:
@RV-Adrian if you don't mind, can you just push this same change to another branch, I'll clean this one up. Then we can make a PR. Thanks.

ghislaineguerin commented on src/scss/ui-components/progress-indicator.scss:
```diff
@@ -0,0 +1,69 @@
+.progress {
```
@RV-Adrian that makes sense.

RV-Adrian replied:
@ghislaineguerin thanks 👍

RV-Adrian commented:
@alex-deaconu Sure, that works 👍

olegrise commented on src/templates/components/purchase-flow/checkout-review-subscription.html:
```diff
@@ -0,0 +1,75 @@
+      <div class="text-center u_padding-md">
+        <p>${{plan.monthly.priceDisplayMonth}}</p>
+        <span>{{ 'common-header.purchase.review-subscription.per-display' | translate}} <br>
+          {{ 'common-header.purchase.review-subscription.per-month' | translate}}
```
@alex-deaconu always "per month" regardless of selection?

alex-deaconu replied:
@olegrise hmm, good question. I assumed it's always monthly because of the prototype.

@RV-Adrian @ezequielc should this show yearly prices if yearly is selected in the Plans Modal or in the Radio selector below? 

Thanks.

RV-Adrian replied:
@alex-deaconu @ezequielc 
Oh yes, the Yearly/Monthly selection following through was the intention. Thx!

alex-deaconu replied:
@olegrise @RV-Adrian okay I will update. Although this was not very clear at all. Thanks.

alex-deaconu replied:
<img width="564" alt="screen shot 2018-07-18 at 5 29 32 pm" src="https://user-images.githubusercontent.com/6218188/42908893-2ea74b28-8ab0-11e8-97c2-1459549055a3.png">

@olegrise @RV-Adrian this looks very weird and confusing (showing yearly values). It's also not consistent with our Plans modal where we keep the same price per Display per Month. Let's discuss further at stand-up tomorrow. @ezequielc FYI. Thanks.

RV-Adrian replied:
@alex-deaconu Ok, I see. I think there was a miscommunication here. I'll break it down and let's chat tomorrow. Thanks!

ezequielc replied:
@alex-deaconu @olegrise @RV-Adrian I see how it can be confusing, but the original implementation was better/correct, in my opinion. We should always show per display per month, the same way as in Plans modal. 
They will see the difference in price by checking the Total below. 
Let's review further tomorrow. Thanks

alex-deaconu replied:
@olegrise @RV-Adrian just want to confirm that as per discussions at stand-up, we're leaving this as is. Thanks.

RV-Adrian replied:
@alex-deaconu Sounds good to me. All other questions that came out of this can be handled on the card I think. 👍

olegrise replied:
👍

olegrise commented on src/templates/components/purchase-flow/purchase-modal.html:
```diff
@@ -0,0 +1,41 @@
+  <div id="purchase-modal" class="modal-body checkout-modal" stop-event="touchend">
+
+    <div id="prototypeProgressIndicator" class="progress-indicator">
+      <div ng-repeat-start="step in PURCHASE_STEPS" class="indicator" ng-class="{ active: currentStep === step.index, complete: currentStep > step.index }" ng-click="setCurrentStep(step)">
```
@alex-deaconu don't forget to remove `ng-click` here.

alex-deaconu replied:
@olegrise I'm allowing navigation from this component, based on the prototype. Are we not supposed to? I guess we can always disable...

olegrise replied:
@alex-deaconu the navigation here is for prototype/debug purposes only. It should not be in the final version. @RV-Adrian you concur?

RV-Adrian replied:
@olegrise is right, the intention of the nav behaviour was to allow movement through the prototype only. In production this is intended to be a progress indicator for the user. @alex-deaconu

alex-deaconu replied:
@olegrise @RV-Adrian okay, I will remove. Thanks.

alex-deaconu replied:
Navigation from the progress indicator was removed. Please see latest commit. Thanks.

alex-deaconu commented:
@Rise-Vision/apps are we good to merge this PR at this point? Thanks.
